layout/plan.h

```diff
@@ -0,0 +1,619 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2016-2021 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ * Original author: Nikita Danilov <nikita.danilov@seagate.com>
+ * Original creation date: 07-Jun-2016
+ */
+
+#pragma once
+
+#ifndef __MOTR_LAYOUT_PLAN_H__
+#define __MOTR_LAYOUT_PLAN_H__
+
+/**
+ * @defgroup layout
+ *
+ * Layout access plan, theory of operation
+ * =======================================
+ *
+ * Overview
+ * --------
+ *
+ * Layout access plan is an abstraction, which describes how a particular
+ * operation is to be executed on an object or index.
+ *
+ * An access plan can be thought of as a directed graph, where vertices are
+ * "plan operations" (plops) and arcs correspond to dependencies between plops.
+ *
+ * Take a network parity de-clustered layout as an example. To read an extent
+ * from an object with this layout, one has to read data from the cobs of the
+ * object. The plan for the object read would, therefore, contain a plop for
+ * each cob read operation, each having an outgoing arc to the "operation
+ * complete" plop.
+ *
+ * In case of a write to an object with parity de-clustered layout, the plan
+ * contains additional plops corresponding to parity calculations. In case of
+ * partial group write, the plan, in addition, contains read-plops describing
+ * the logic of read-modify-write cycle. Degraded mode introduces more plops and
+ * so on.
+ *
+ * Access plans are an intermediate abstraction designed for two main users:
+ *
+ *   - IO code in client API, including key-value indices and data objects, and
```

Comment:
Do you mean, when the "Access plan" APIs are ready, we need some kind of re-write of client code?

Reply:
As told in the introduction PDF, client i/o code may be re-written to use these APIs.

Reply:
Yes, it can be rewritten later when the plan API implementation is ready and stable.